Ensure that all spark tools apply the same read filters as their walker equivalents

droazen opened the pull request:
Most ReadWalkers apply the WellformedReadFilter, but their spark equivalents
were not doing so. This creates an inherited method GATKSparkTool.makeReadFilter()
that defaults to the WellformedReadFilter and gets automatically applied to the
RDD of reads returned from GATKSparkTool.getReads().

Only BaseRecalibratorSpark needed to override this method to apply a custom
read filter in order to match the walker filtering settings.

Resolves #1158

droazen commented:
For @akiezun or @lbergelson to review (whoever is around)

droazen commented:
Forgot to modify the multi-tool pipelines like `BQSRPipelineSpark` -- will do this now.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortBamSpark.java:
```diff
@@ -32,6 +34,11 @@
 
     @Override
+    public CountingReadFilter makeReadFilter() {
+        return new CountingReadFilter("Allow all reads", ReadFilterLibrary.ALLOW_ALL_READS);
```
Does it make sense to use a counting read filter with spark tools?

cmnbroad replied:
I was just going to make the same comment. They're stateful, so I would
think not.

On Thu, Nov 19, 2015 at 2:15 PM, Louis Bergelson <notifications@github.com>
wrote:

> In
> src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortBamSpark.java
> <https://github.com/broadinstitute/gatk/pull/1159#discussion_r45386161>:
>
> > @@ -32,6 +34,11 @@
> >      protected int parallelism = 0;
> >
> >      @Override
> > +    public CountingReadFilter makeReadFilter() {
> > +        return new CountingReadFilter("Allow all reads", ReadFilterLibrary.ALLOW_ALL_READS);
>
> Does it make sense to use a counting read filter with spark tools?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1159/files#r45386161>.
>

droazen replied:
Fair point, I will switch to using raw ReadFilters then for spark.

droazen commented:
Ok, I've modified the multi-tool pipelines to apply the right read filters for each tool in the pipeline. I've also modified the spark metrics tools to not do any filtering at all.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java:
```diff
@@ -197,6 +216,17 @@ public SAMFileHeader getHeaderForReads() {
+     * Subclasses can extend to provide their own filters (ie., override and optionally call super).
+     * Multiple filters can be composed by using {@link org.broadinstitute.hellbender.engine.filters.ReadFilter} composition methods.
+     */
+    public CountingReadFilter makeReadFilter() {
```
Should we just use a regular read filter instead of a counting one?  It seems like minor but unnecessary overhead since the counts won't be used in any way?

droazen replied:
Ok, will change.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/BQSRPipelineSpark.java:
```diff
@@ -90,13 +94,20 @@ protected void runTool(final JavaSparkContext ctx) {
+        // The initial reads have already had the WellformedReadFilter applied to them, which
+        // is all the filtering that ApplyBQSR wants. BQSR itself wants additional filtering
+        // performed, so we do that here.
+        final ReadFilter bqsrReadFilter = BaseRecalibrator.makeBQSRSpecificReadFilters();
```
It seems worse to have to pull this out rather than having it in the tool itself.

droazen replied:
Will be fixed in beta :)

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java:
```diff
@@ -176,12 +179,28 @@ public SAMFileHeader getHeaderForReads() {
      */
     public JavaRDD<GATKRead> getReads() {
+        final ReadFilter filter = makeReadFilter();
+        return getUnfilteredReads().filter(read -> filter.apply(read));
```
tiny micro optimization that probably doesn't matter at all becauser of JIT, but you should prefer `filter.test` over `filter.apply`

droazen replied:
Fixed

lbergelson commented:
@droazen I'm not sure this is an improvement.  We want the fundamental unit of spark tool to be the transform, not the cli wrapper around it.  If we do this then we're pushing more of the contract of the transform outside of itself, i.e. see the newly duplicated bqsr code.  I think that it was a deliberate decision to lift all reads into the initial rdd and then filter them in the transforms to what was needed by that transform.  This is paying some performance cost in multi-stage pipelines which will potentially apply the same filters over and over again, but it simplifies the code because the filters can be baked into the transform and the pipeline writer doesn't have to think about them.

It would be nice if we had a mechanism for adding metadata to an RDD so we can say "this is a sorted RDD filtered with X,Y,and Z filters", so we could intelligently avoid re-filtering.

droazen commented:
@lbergelson It's an improvement in the sense that we had a bug where the spark tools were not doing the same filtering as their walker equivalents, and this is a quick fix for the bug that will also help avoid the problem in the future.

I agree that the current design is problematic for multi-tool pipelines, but do you think we could address this in a separate ticket after alpha, and live with an imperfect design for now?

akiezun commented:
I agree with @lbergelson that we need to improved this long term and be able to compose transforms in a nice way. But I also agree with @droazen that the current state is a bug which needs to be fixed. Let's fix the bug for alpha and improve the world for beta. WDYT?

akiezun commented:
btw, @droazen can you add 1 test at least that fails without your change and works with the change? To demonstrate that the change does something and to prevent regressions.

droazen commented:
Yeah I think we need a true Transform abstraction before we can do what @lbergelson wants and not duplicate code everywhere, and this seems like a beta addition to me.

lbergelson commented:
If this fixes an immediate problem I'm fine merging it.  I think we should revisit though for beta.

droazen commented:
Ok, I've added a test for `PrintReadsSpark` that fails without the filtering, and passes with it on, and addressed the other (non-design) comments. Will merge once tests pass, and create a new beta ticket to address the design concerns.
